// rosaenlg-commons/dist/LanguageCommonItalian.js
"use strict";
/**
 * @license
 * Copyright 2019 Ludan Stoecklé
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LanguageCommonItalian = void 0;
const LanguageCommon_1 = require("./LanguageCommon");
class LanguageCommonItalian extends LanguageCommon_1.LanguageCommon {
    constructor() {
        super(...arguments);
        this.iso2 = 'it';
        this.validPropsWord = ['G', 'S', 'P'];
        this.validPropsAdj = ['MS', 'MP', 'FS', 'FP'];
    }
    isConsonneImpure(word) {
        const wordLc = word.toLowerCase();
        const begins = ['ps', 'pn', 'gn', 'x', 'z'];
        for (let i = 0; i < begins.length; i++) {
            //console.log(`${word} start with ${begins[i]}?`);
            if (wordLc.startsWith(begins[i])) {
                //console.log(`isConsonneImpure ${word}? => true`);
                return true;
            }
        }
        // s impur (autrement dit un s suivi d'une autre consonne)
        const regexSImpur = new RegExp('^s[' + this.constants.toutesConsonnes + ']');
        if (regexSImpur.test(wordLc)) {
            //console.log(`isConsonneImpure ${word}? => true`);
            return true;
        }
        //console.log(`isConsonneImpure ${word}? => false`);
        return false;
    }
    isIFollowedByVowel(word) {
        const regexISuiviVoyelle = new RegExp('^[IiYy][' + this.constants.toutesVoyellesMinuscules + ']');
        if (regexISuiviVoyelle.test(word)) {
            return true;
        }
        return false;
    }
    startsWithVowel(word) {
        const regexVowel = new RegExp('^[' + this.constants.toutesVoyellesMinuscules + ']');
        if (regexVowel.test(word.toLowerCase())) {
            return true;
        }
        return false;
    }
}
exports.LanguageCommonItalian = LanguageCommonItalian;
